hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java

```diff
@@ -225,52 +228,71 @@ public void setHederaLedger(final HederaLedger hederaLedger) {
-				validity = TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED;
-			} else {
-				var maxAutomaticAssociations = (int) accountsLedger.get(aId, MAX_AUTOMATIC_ASSOCIATIONS);
-				var alreadyUsedAutomaticAssociations = (int) accountsLedger.get(aId,
-						ALREADY_USED_AUTOMATIC_ASSOCIATIONS);
+			var maxAutomaticAssociations = (int) accountsLedger.get(aId, MAX_AUTOMATIC_ASSOCIATIONS);
+			var alreadyUsedAutomaticAssociations = (int) accountsLedger.get(aId,
+					ALREADY_USED_AUTOMATIC_ASSOCIATIONS);
 
-				if (automaticAssociation && alreadyUsedAutomaticAssociations >= maxAutomaticAssociations) {
-					validity = NO_REMAINING_AUTOMATIC_ASSOCIATIONS;
-				}
+			if (automaticAssociation && alreadyUsedAutomaticAssociations >= maxAutomaticAssociations) {
+				validity = NO_REMAINING_AUTOMATIC_ASSOCIATIONS;
+			}
 
-				if (validity == OK) {
-					accountTokens.associateAll(new HashSet<>(tokenIds));
-					for (var id : tokenIds) {
-						final var relationship = asTokenRel(aId, id);
-						tokenRelsLedger.create(relationship);
-						final var token = get(id);
-						tokenRelsLedger.set(
-								relationship,
-								TokenRelProperty.IS_FROZEN,
-								token.hasFreezeKey() && token.accountsAreFrozenByDefault());
-						tokenRelsLedger.set(
-								relationship,
-								TokenRelProperty.IS_KYC_GRANTED,
-								!token.hasKycKey());
-						tokenRelsLedger.set(
-								relationship,
-								TokenRelProperty.IS_AUTOMATIC_ASSOCIATION,
-								automaticAssociation);
-
-						sideEffectsTracker.trackAutoAssociation(id, aId);
-						if (automaticAssociation) {
-							accountsLedger.set(aId, ALREADY_USED_AUTOMATIC_ASSOCIATIONS,
-									alreadyUsedAutomaticAssociations + 1);
-						}
+			if (validity == OK) {
+				final var tokenAssociationMetadata =
+						(TokenAssociationMetadata) accountsLedger.get(aId, TOKEN_ASSOCIATION_METADATA);
+				final var lastAssociatedToken = tokenAssociationMetadata.lastAssociation();
+				var numAssociations = tokenAssociationMetadata.numAssociations();
+				var numZeroBalances = tokenAssociationMetadata.numZeroBalances();
+				var currKey = new EntityNumPair(lastAssociatedToken.value());
+				for (var id : tokenIds) {
+					final var relationship = asTokenRel(aId, id);
+					tokenRelsLedger.create(relationship);
+					final var token = get(id);
+					tokenRelsLedger.set(
+							relationship,
+							TokenRelProperty.IS_FROZEN,
+							token.hasFreezeKey() && token.accountsAreFrozenByDefault());
+					tokenRelsLedger.set(
+							relationship,
+							TokenRelProperty.IS_KYC_GRANTED,
+							!token.hasKycKey());
+					tokenRelsLedger.set(
+							relationship,
+							TokenRelProperty.IS_AUTOMATIC_ASSOCIATION,
+							automaticAssociation);
+
+					sideEffectsTracker.trackAutoAssociation(id, aId);
+					if (automaticAssociation) {
+						accountsLedger.set(aId, ALREADY_USED_AUTOMATIC_ASSOCIATIONS,
+								alreadyUsedAutomaticAssociations + 1);
```

Comment:
This would be a bug if this method was ever used with a list of more than one `TokenID` (b/c we would only increase `alreadyUsedAutomaticAssociations` by at most 1, no matter how many automatic associations were done).

Luckily, we _do_ only call this [with one `TokenID`](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java#L900); so imo we should simplify by removing the `List<TokenID>` in the signature, replacing with just `TokenID`.